scripts/Diffraction/isis_powder/hrpd.py

```diff
@@ -235,71 +235,37 @@ def _subtract_prompt_pulses(self, ws, tof_res=0.002):
                 cen = npulse * PROMPT_PULSE_INTERVAL
                 xlo = cen - PROMPT_PULSE_LEFT_WIDTH - 10  # add extra to ensure get background on both sides of peak
                 xhi = cen + PROMPT_PULSE_RIGHT_WIDTH + 10
-                comp_func = FunctionFactory.createInitialized(
-                    f"name=PearsonIV, Centre={8}, Intensity=1,Sigma=8.5, Exponent=1.5, Skew=-5,"
-                    f"constraints=(0.2<Sigma,1.5<Exponent);name=FlatBackground, A0=0,constraints=(0<A0)"
-                )
-                comp_func[0].setAttributeValue("CentreShift", cen)
-                comp_func[0].setHeight(ws.readY(ispec)[ws.yIndexOfX(cen)])
-                comp_func[0].addConstraints(f"{-15}<Centre<{15}")
-                comp_func[0].addConstraints(f"{0}<Intensity")
-                comp_func[1]["A0"] = min(ws.readY(ispec)[ws.yIndexOfX(xlo)], ws.readY(ispec)[ws.yIndexOfX(xhi)])
+                comp_func = self._setup_single_prompt_pulse_function(ws, ispec, cen, xlo, xhi)
                 func.add(comp_func)
                 func.setDomainIndex(ipulse, ipulse)
                 key_suffix = f"_{ipulse}" if ipulse > 0 else ""
                 fit_kwargs["InputWorkspace" + key_suffix] = ws.name()
                 fit_kwargs["StartX" + key_suffix] = xlo
                 fit_kwargs["EndX" + key_suffix] = xhi
                 fit_kwargs["WorkspaceIndex" + key_suffix] = ispec
-                # work out how to exclude
-                exclude = []
-                diff_consts = si.diffractometerConstants(ispec)
-                for dpk in dpks_bragg:
-                    tof_pk = UnitConversion.run("dSpacing", "TOF", dpk, 0, DeltaEModeType.Elastic, diff_consts)
-                    tof_pk_lo = tof_pk * (1 - tof_res)
-                    tof_pk_hi = tof_pk * (1 + tof_res)
-                    if xlo < tof_pk_lo < xhi or xlo < tof_pk_hi < xhi or (tof_pk_lo < xlo and tof_pk_hi > xhi):
-                        exclude.extend([tof_pk_lo, tof_pk_hi])
+                # exclude x-ranges where Bragg peaks overlap prompt pulse within some fractional resolution
+                exclude = self._get_tofs_to_exclude(si, ispec, dpks_bragg, xlo, xhi)
                 if exclude:
                     func.fixParameter(f"f{ipulse}.f1.A0")  # fix constant background - Bragg peak can mess with bg
                     fit_kwargs["Exclude" + key_suffix] = exclude
             # check that at least one fit region has no overlapping prompt pulse
             if len([key for key in fit_kwargs.keys() if "Exclude" in key]) < len(npulses):
```

Comment:
Sorry, a small nitpick, I think it is better to extract this list because the condition is too long and confusing
```suggestion
          excluded_keys = [key for key in fit_kwargs.keys() if "Exclude" in key]
          if len(excluded_keys) < len(npulses):
```